README.md
# lazy_data_table_plus

A Flutter widget data table that can be loaded lazily. The table also has a column header row and a row header column that will stay in view.
(This widget is still in development, and may not work 100%)

This widget is based on [table-sticky-headers](https://pub.dev/packages/table_sticky_headers) made by Alex Bacich, so credits to him.

## Features

* Scrollable when items overflow
* Column header row stays in view
* Row header column stays in view
* Items are loaded lazily

![img not loaded](https://github.com/aguilaair/lazy_data_table_plus/raw/master/example/lazy_data_table_plus_example.gif "lazy_data_table_plus example")

## Usage

To use this widget, add `lazy_data_table_plus: ^0.2.0` to your dependencies in `pubspec.yaml`

```yaml
dependencies:
  lazy_data_table_plus: ^0.2.0
```

Then the package can be included in a file with:

```dart
import 'package:lazy_data_table_plus/lazy_data_table_plus.dart';
```

And then the LazyDataTable can be used as following:  
(This example is used to create the table in the gif above)

```dart
LazyDataTable(
  rows: 100,
  columns: 100,
  tableDimensions: LazyDataTableDimensions(
    cellHeight: 50,
    cellWidth: 100,
    columnHeaderHeight: 50,
    rowHeaderWidth: 75,
  ),
  tableTheme: LazyDataTableTheme(
    columnHeaderBorder: Border.all(color: Colors.black38),
    rowHeaderBorder: Border.all(color: Colors.black38),
    cellBorder: Border.all(color: Colors.black12),
    cornerBorder: Border.all(color: Colors.black38),
    columnHeaderColor: Colors.white60,
    rowHeaderColor: Colors.white60,
    cellColor: Colors.white,
    cornerColor: Colors.white38,
  ),
  columnHeaderBuilder: (i) => Center(child: Text("Column: ${i + 1}")),
  rowHeaderBuilder: (i) => Center(child: Text("Row: ${i + 1}")),
  dataCellBuilder: (i, j) => Center(child: Text("Cell: $i, $j")),
  cornerWidget: Center(child: Text("Corner")),
),
```

## Issues
If you have any problems or even suggestions feel free leave them [here](https://github.com/aguilaair/lazy_data_table_plus/issues)


CHANGELOG.md
## [1.0.0]

* Support null safety
## [0.2.1]

* Forgot to update docs!

## [0.2.0]

This release brings lazy_data_table_plus on-par with lazy_data_table after several fixes from the author!

* Updated docs to new dependency.
* Fix Scroll issue (Fixed by Naomi on lazy_data_table)
* Change TableDimensions to LazyDataTableDimensions (Chnaged by Naomi on lazy_data_table)

## [0.1.6]

* Fixed bug that stopped themeing from happening.

## [0.1.5]

* Fixed bug that caused headers to scroll separately.
* Expanded example.

## [0.1.4]

* Changed NotificationListener to GestureDetector.
* This allows for the ability to scroll in both directions at the same time.(diagonally)

## [0.1.3]

* Added DataTableDimensions, which holds information about the dimensions of the table.
* Added DataTableTheme, which holds information about the colors and borders of the table.
* Added example gif to the README.

## [0.1.2]

* Updated synchronized ScrollController to be more efficient.

## [0.1.1]

* Column and row headers are now optional.

## [0.0.1]

* Initial release


test/lazy_data_table_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:lazy_data_table_plus/lazy_data_table_plus.dart';

void main() {}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'headerless_table.dart';
import 'simple_table.dart';

void main() {
  runApp(MaterialApp(
    title: 'Example',
    home: MyApp(),
  ));
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Examples")),
      body: Center(
        child: Column(
          children: <Widget>[
            ElevatedButton(
              child: Text("Simple table"),
              onPressed: () => Navigator.push(context,
                  MaterialPageRoute(builder: (context) => SimpleTable())),
            ),
            ElevatedButton(
              child: Text("Header-less table"),
              onPressed: () => Navigator.push(context,
                  MaterialPageRoute(builder: (context) => HeaderlessTable())),
            ),
          ],
        ),
      ),
    );
  }
}


lib/lazy_data_table.dart
library lazy_data_table;

import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';
import 'package:flutter/scheduler.dart';

/// Create a lazily loaded data table.
///
/// The table is [columns] by [rows] big.
/// The [columnHeaderBuilder] and [rowHeaderBuilder] are optional,
/// and when either of those is not given, the corner widget should also not be given.
class LazyDataTable extends StatefulWidget {
  LazyDataTable({
    Key key,
    // Number of data columns.
    @required this.columns,

    // Number of data rows.
    @required this.rows,

    // Dimensions of the table elements.
    this.tableDimensions = const DataTableDimensions(),

    // Theme of the table elements.
    this.tableTheme = const DataTableTheme(),

    // Builder function for the column header.
    this.columnHeaderBuilder,

    // Builder function for the row header.
    this.rowHeaderBuilder,

    // Builder function for the data cell.
    @required this.dataCellBuilder,

    // Corner widget.
    this.cornerWidget,
  }) : super(key: key) {
    assert(columns != null);
    assert(rows != null);
    assert(dataCellBuilder != null);
    if (rowHeaderBuilder == null || columnHeaderBuilder == null) {
      assert(cornerWidget == null,
          "The corner widget is only allowed when you have both a column header and a row header.");
    }
  }

  /// The state class that contains the table.
  final table = _LazyDataTableState();

  // Amount of cells
  /// The number of columns in the table.
  final int columns;

  /// The number of rows in the table.
  final int rows;

  // Size of cells and headers
  /// The dimensions of the table cells and headers.
  final DataTableDimensions tableDimensions;

  // Theme of the table
  /// The theme of the table cells and headers.
  final DataTableTheme tableTheme;

  // Builder functions
  /// The builder function for a column header.
  final Widget Function(int columnIndex) columnHeaderBuilder;

  /// The builder function for a row header.
  final Widget Function(int rowIndex) rowHeaderBuilder;

  /// The builder function for a data cell.
  final Widget Function(int columnIndex, int rowIndex) dataCellBuilder;

  /// The widget for the upper-left corner.
  final Widget cornerWidget;

  @override
  _LazyDataTableState createState() => table;

  /// Jump the table to the given cell.
  void jumpToCell(int column, int row) {
    table.jumpToCell(column, row);
  }

  /// Jump the table to the given location.
  void jumpTo(double x, double y) {
    table.jumpTo(x, y);
  }
}

class _LazyDataTableState extends State<LazyDataTable>
    with TickerProviderStateMixin {
  _CustomScrollController _horizontalControllers;
  _CustomScrollController _verticalControllers;

  @override
  void initState() {
    super.initState();

    _horizontalControllers = _CustomScrollController(this);
    _verticalControllers = _CustomScrollController(this);
  }

  @override
  void dispose() {
    _horizontalControllers.dispose();
    _verticalControllers.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanUpdate: (DragUpdateDetails details) {
        jump(-details.delta.dx, -details.delta.dy);
      },
      onPanEnd: (DragEndDetails details) {
        _verticalControllers
            .setVelocity(-details.velocity.pixelsPerSecond.dy / 100);
        _horizontalControllers
            .setVelocity(-details.velocity.pixelsPerSecond.dx / 100);
      },
      child: Row(
        children: <Widget>[
          widget.rowHeaderBuilder != null
              ? SizedBox(
                  width: widget.tableDimensions.rowHeaderWidth,
                  child: Column(
                    children: <Widget>[
                      // Corner widget
                      SizedBox(
                        height: widget.tableDimensions.columnHeaderHeight,
                        width: widget.tableDimensions.rowHeaderWidth,
                        child: widget.cornerWidget != null
                            ? Container(
                                decoration: BoxDecoration(
                                  color: widget.tableTheme.cornerColor,
                                  border: widget.tableTheme.cornerBorder,
                                ),
                                child: widget.cornerWidget,
                              )
                            : Container(),
                      ),
                      // Row headers
                      Expanded(
                        child: ListView.builder(
                            scrollDirection: Axis.vertical,
                            controller: _verticalControllers,
                            physics: NeverScrollableScrollPhysics(),
                            itemCount: widget.rows,
                            itemBuilder: (__, i) {
                              return Container(
                                height: widget.tableDimensions.cellHeight,
                                width: widget.tableDimensions.rowHeaderWidth,
                                decoration: BoxDecoration(
                                  color: widget.tableTheme.rowHeaderColor,
                                  border: widget.tableTheme.rowHeaderBorder,
                                ),
                                child: widget.rowHeaderBuilder(i),
                              );
                            }),
                      ),
                    ],
                  ),
                )
              : Container(),
          Expanded(
            child: Column(
              children: <Widget>[
                // Column headers
                widget.columnHeaderBuilder != null
                    ? SizedBox(
                        height: widget.tableDimensions.columnHeaderHeight,
                        child: ListView.builder(
                            scrollDirection: Axis.horizontal,
                            controller: _horizontalControllers,
                            physics: NeverScrollableScrollPhysics(),
                            itemCount: widget.columns,
                            itemBuilder: (__, i) {
                              return Container(
                                height:
                                    widget.tableDimensions.columnHeaderHeight,
                                width: widget.tableDimensions.cellWidth,
                                decoration: BoxDecoration(
                                  color: widget.tableTheme.columnHeaderColor,
                                  border: widget.tableTheme.columnHeaderBorder,
                                ),
                                child: widget.columnHeaderBuilder(i),
                              );
                            }),
                      )
                    : Container(),
                // Main data
                Expanded(
                  // List of rows
                  child: ListView.builder(
                      scrollDirection: Axis.vertical,
                      shrinkWrap: true,
                      controller: _verticalControllers,
                      physics: NeverScrollableScrollPhysics(),
                      itemCount: widget.rows,
                      itemBuilder: (_, i) {
                        // Single row
                        return SizedBox(
                          height: widget.tableDimensions.cellHeight,
                          child: ListView.builder(
                              scrollDirection: Axis.horizontal,
                              shrinkWrap: true,
                              controller: _horizontalControllers,
                              physics: NeverScrollableScrollPhysics(),
                              itemCount: widget.columns,
                              itemBuilder: (__, j) {
                                return Container(
                                  height: widget.tableDimensions.cellHeight,
                                  width: widget.tableDimensions.cellWidth,
                                  decoration: BoxDecoration(
                                    color: widget.tableTheme.cellColor,
                                    border: widget.tableTheme.cellBorder,
                                  ),
                                  child: widget.dataCellBuilder(i, j),
                                );
                              }),
                        );
                      }),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  /// Jump the table to the given cell.
  void jumpToCell(int column, int row) {
    _horizontalControllers.jumpTo(column * widget.tableDimensions.cellWidth);
    _verticalControllers.jumpTo(row * widget.tableDimensions.cellHeight);
  }

  /// Jump the table to the given location.
  void jumpTo(double x, double y) {
    _horizontalControllers.jumpTo(x);
    _verticalControllers.jumpTo(y);
  }

  /// Jump to a relative location from the current location.
  void jump(double x, double y) {
    _horizontalControllers.jump(x);
    _verticalControllers.jump(y);
  }
}

/// Data class for the dimensions of a [LazyDataTable].
class DataTableDimensions {
  const DataTableDimensions({
    this.cellHeight = 50,
    this.cellWidth = 50,
    this.columnHeaderHeight = 50,
    this.rowHeaderWidth = 50,
  });

  /// Height of a cell and row header.
  final double cellHeight;

  /// Width of a cell and column header.
  final double cellWidth;

  /// Height of a column header.
  final double columnHeaderHeight;

  /// Width of a row header.
  final double rowHeaderWidth;
}

/// Data class for the theme of a [LazyDataTable].
class DataTableTheme {
  const DataTableTheme({
    this.columnHeaderBorder,
    this.rowHeaderBorder,
    this.cellBorder,
    this.cornerBorder,
    this.columnHeaderColor,
    this.rowHeaderColor,
    this.cellColor,
    this.cornerColor,
  });

  /// [BoxBorder] for the column header.
  final BoxBorder columnHeaderBorder;

  /// [BoxBorder] for the row header.
  final BoxBorder rowHeaderBorder;

  /// [BoxBorder] for the cell.
  final BoxBorder cellBorder;

  /// [BoxBorder] for the corner widget.
  final BoxBorder cornerBorder;

  /// [Color] for the column header.
  final Color columnHeaderColor;

  /// [Color] for the row header.
  final Color rowHeaderColor;

  /// [Color] for the cell.
  final Color cellColor;

  /// [Color] for the corner widget.
  final Color cornerColor;
}

/// A custom synchronized scroll controller.
///
/// This controller stores all their attached [ScrollPosition] in a list,
/// and when given a notification via [processNotification], it will scroll
/// every ScrollPosition in that list to the same [offset].
class _CustomScrollController extends ScrollController {
  _CustomScrollController(TickerProvider provider) : super() {
    _ticker = provider.createTicker((_) {
      jumpTo(offset + _velocity);
      _velocity *= 0.9;
      if (_velocity < 0.1 && _velocity > -0.1) {
        _ticker.stop();
      }
    });
  }

  /// List of [ScrollPosition].
  List<ScrollPosition> _positions = List();

  /// The offset of the ScrollPositions.
  double offset = 0;

  /// Ticker to calculate the [_velocity].
  Ticker _ticker;

  /// The velocity of the controller.
  /// The [_ticker] will tick while the velocity
  /// is not between -0.1 and 0.1.
  double _velocity;

  /// Stores given [ScrollPosition] in the list and
  /// set the initial offset of that ScrollPosition.
  @override
  void attach(ScrollPosition position) {
    position.correctPixels(offset);
    _positions.add(position);
  }

  /// Removes given [ScrollPostion] from the list.
  @override
  void detach(ScrollPosition position) {
    _positions.remove(position);
  }

  /// Processes notification from one of the [ScrollPositions] in the list.
  void processNotification(ScrollNotification notification) {
    if (notification is ScrollUpdateNotification) {
      jumpTo(notification.metrics.pixels);
    }
  }

  /// Jumps every item in the list to the given [value],
  /// except the ones that are already at the correct offset.
  @override
  void jumpTo(double value) {
    if (_positions[0] != null && value > _positions[0].maxScrollExtent) {
      offset = _positions[0].maxScrollExtent;
    } else if (value < 0) {
      offset = 0;
    } else {
      offset = value;
    }
    for (ScrollPosition position in _positions) {
      if (position.pixels != offset) {
        position.jumpTo(offset);
      }
    }
  }

  /// Jump to [offset] + [value].
  void jump(double value) {
    jumpTo(offset + value);
  }

  /// Set [_velocity] to new value.
  void setVelocity(double velocity) {
    if (_ticker.isActive) _ticker.stop();
    _velocity = velocity;
    _ticker.start();
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }
}


lib/lazy_data_table_plus.dart
library lazy_data_table;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';
import 'package:flutter/scheduler.dart';

/// Create a lazily loaded data table.
///
/// The table is [columns] by [rows] big.
/// The [columnHeaderBuilder] and [rowHeaderBuilder] are optional,
/// and when either of those is not given, the corner widget should also not be given.
class LazyDataTable extends StatefulWidget {
  LazyDataTable({
    Key? key,
    // Number of data columns.
    required this.columns,

    // Number of data rows.
    required this.rows,

    // Dimensions of the table elements.
    this.tableDimensions = const LazyDataTableDimensions(),

    // Theme of the table elements.
    this.tableTheme = const LazyDataTableTheme(),

    // Builder function for the column header.
    this.columnHeaderBuilder,

    // Builder function for the row header.
    this.rowHeaderBuilder,

    // Builder function for the data cell.
    required this.dataCellBuilder,

    // Corner widget.
    this.cornerWidget,
  }) : super(key: key) {
    if (rowHeaderBuilder == null || columnHeaderBuilder == null) {
      assert(cornerWidget == null,
          "The corner widget is only allowed when you have both a column header and a row header.");
    }
  }

  /// The state class that contains the table.
  final table = _LazyDataTableState();

  // Amount of cells
  /// The number of columns in the table.
  final int columns;

  /// The number of rows in the table.
  final int rows;

  // Size of cells and headers
  /// The dimensions of the table cells and headers.
  final LazyDataTableDimensions tableDimensions;

  // Theme of the table
  /// The theme of the table cells and headers.
  final LazyDataTableTheme tableTheme;

  // Builder functions
  /// The builder function for a column header.
  final Widget Function(int columnIndex)? columnHeaderBuilder;

  /// The builder function for a row header.
  final Widget Function(int rowIndex)? rowHeaderBuilder;

  /// The builder function for a data cell.
  final Widget Function(int columnIndex, int rowIndex) dataCellBuilder;

  /// The widget for the upper-left corner.
  final Widget? cornerWidget;

  @override
  _LazyDataTableState createState() => table;

  /// Jump the table to the given cell.
  void jumpToCell(int column, int row) {
    table.jumpToCell(column, row);
  }

  /// Jump the table to the given location.
  void jumpTo(double x, double y) {
    table.jumpTo(x, y);
  }
}

class _LazyDataTableState extends State<LazyDataTable>
    with TickerProviderStateMixin {
  _CustomScrollController? _horizontalControllers;
  _CustomScrollController? _verticalControllers;

  @override
  void initState() {
    super.initState();

    _horizontalControllers = _CustomScrollController(this);
    _verticalControllers = _CustomScrollController(this);
  }

  @override
  void dispose() {
    _horizontalControllers!.dispose();
    _verticalControllers!.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Listener(
      onPointerSignal: (pointerSignal) {
        if (pointerSignal is PointerScrollEvent) {
          jump(pointerSignal.scrollDelta.dx, pointerSignal.scrollDelta.dy);
        }
      },
      child: GestureDetector(
        onPanUpdate: (DragUpdateDetails details) {
          jump(-details.delta.dx, -details.delta.dy);
        },
        onPanEnd: (DragEndDetails details) {
          _verticalControllers!
              .setVelocity(-details.velocity.pixelsPerSecond.dy / 100);
          _horizontalControllers!
              .setVelocity(-details.velocity.pixelsPerSecond.dx / 100);
        },
        child: Row(
          children: <Widget>[
            widget.rowHeaderBuilder != null
                ? SizedBox(
                    width: widget.tableDimensions.rowHeaderWidth,
                    child: Column(
                      children: <Widget>[
                        // Corner widget
                        SizedBox(
                          height: widget.tableDimensions.columnHeaderHeight,
                          width: widget.tableDimensions.rowHeaderWidth,
                          child: widget.cornerWidget != null
                              ? Container(
                                  decoration: BoxDecoration(
                                    color: widget.tableTheme.cornerColor,
                                    border: widget.tableTheme.cornerBorder,
                                  ),
                                  child: widget.cornerWidget,
                                )
                              : Container(),
                        ),
                        // Row headers
                        Expanded(
                          child: ListView.builder(
                              scrollDirection: Axis.vertical,
                              controller: _verticalControllers,
                              physics: NeverScrollableScrollPhysics(),
                              itemCount: widget.rows,
                              itemBuilder: (__, i) {
                                return Container(
                                  height: widget
                                          .tableDimensions.customCellHeight
                                          .containsKey(i)
                                      ? widget
                                          .tableDimensions.customCellHeight[i]
                                      : widget.tableDimensions.cellHeight,
                                  width: widget.tableDimensions.rowHeaderWidth,
                                  decoration: BoxDecoration(
                                    color: widget.tableTheme.rowHeaderColor,
                                    border: widget.tableTheme.rowHeaderBorder,
                                  ),
                                  child: widget.rowHeaderBuilder!(i),
                                );
                              }),
                        ),
                      ],
                    ),
                  )
                : Container(),
            Expanded(
              child: Column(
                children: <Widget>[
                  // Column headers
                  widget.columnHeaderBuilder != null
                      ? SizedBox(
                          height: widget.tableDimensions.columnHeaderHeight,
                          child: ListView.builder(
                              scrollDirection: Axis.horizontal,
                              controller: _horizontalControllers,
                              physics: NeverScrollableScrollPhysics(),
                              itemCount: widget.columns,
                              itemBuilder: (__, i) {
                                return Container(
                                  height:
                                      widget.tableDimensions.columnHeaderHeight,
                                  width: widget.tableDimensions.customCellWidth
                                          .containsKey(i)
                                      ? widget
                                          .tableDimensions.customCellWidth[i]
                                      : widget.tableDimensions.cellWidth,
                                  decoration: BoxDecoration(
                                    color: widget.tableTheme.columnHeaderColor,
                                    border:
                                        widget.tableTheme.columnHeaderBorder,
                                  ),
                                  child: widget.columnHeaderBuilder!(i),
                                );
                              }),
                        )
                      : Container(),
                  // Main data
                  Expanded(
                    // List of rows
                    child: ListView.builder(
                        scrollDirection: Axis.vertical,
                        shrinkWrap: true,
                        controller: _verticalControllers,
                        physics: NeverScrollableScrollPhysics(),
                        itemCount: widget.rows,
                        itemBuilder: (_, i) {
                          // Single row
                          return SizedBox(
                            height: widget.tableDimensions.customCellHeight
                                    .containsKey(i)
                                ? widget.tableDimensions.customCellHeight[i]
                                : widget.tableDimensions.cellHeight,
                            child: ListView.builder(
                                scrollDirection: Axis.horizontal,
                                shrinkWrap: true,
                                controller: _horizontalControllers,
                                physics: NeverScrollableScrollPhysics(),
                                itemCount: widget.columns,
                                itemBuilder: (__, j) {
                                  return Container(
                                    height: widget
                                            .tableDimensions.customCellHeight
                                            .containsKey(i)
                                        ? widget
                                            .tableDimensions.customCellHeight[i]
                                        : widget.tableDimensions.cellHeight,
                                    width: widget
                                            .tableDimensions.customCellWidth
                                            .containsKey(j)
                                        ? widget
                                            .tableDimensions.customCellWidth[j]
                                        : widget.tableDimensions.cellWidth,
                                    decoration: BoxDecoration(
                                      color: widget.tableTheme.cellColor,
                                      border: widget.tableTheme.cellBorder,
                                    ),
                                    child: widget.dataCellBuilder(i, j),
                                  );
                                }),
                          );
                        }),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  /// Jump the table to the given cell.
  void jumpToCell(int column, int row) {
    double customWidth = 0;
    int customWidthCells = 0;
    for (int i = 0; i < column; i++) {
      if (widget.tableDimensions.customCellWidth.containsKey(i)) {
        customWidth += widget.tableDimensions.customCellWidth[i]!;
        customWidthCells++;
      }
    }
    _horizontalControllers!.jumpTo(
        (column - customWidthCells) * widget.tableDimensions.cellWidth +
            customWidth);

    double customHeight = 0;
    int customHeightCells = 0;
    for (int i = 0; i < column; i++) {
      if (widget.tableDimensions.customCellHeight.containsKey(i)) {
        customHeight += widget.tableDimensions.customCellHeight[i]!;
        customHeightCells++;
      }
    }
    _verticalControllers!.jumpTo(
        (row - customHeightCells) * widget.tableDimensions.cellHeight +
            customHeight);
  }

  /// Jump the table to the given location.
  void jumpTo(double x, double y) {
    _horizontalControllers!.jumpTo(x);
    _verticalControllers!.jumpTo(y);
  }

  /// Jump to a relative location from the current location.
  void jump(double x, double y) {
    _horizontalControllers!.jump(x);
    _verticalControllers!.jump(y);
  }
}

/// Data class for the dimensions of a [LazyDataTable].
class LazyDataTableDimensions {
  const LazyDataTableDimensions({
    this.cellHeight = 50,
    this.cellWidth = 50,
    this.columnHeaderHeight = 50,
    this.rowHeaderWidth = 50,
    this.customCellHeight = const {},
    this.customCellWidth = const {},
  });

  /// Height of a cell and row header.
  final double cellHeight;

  /// Width of a cell and column header.
  final double cellWidth;

  /// Height of a column header.
  final double columnHeaderHeight;

  /// Width of a row header.
  final double rowHeaderWidth;

  /// Map with the custom height for a certain rows.
  final Map<int, double> customCellHeight;

  /// Map with the custom width for certain columns.
  final Map<int, double> customCellWidth;
}

/// Data class for the theme of a [LazyDataTable].
class LazyDataTableTheme {
  const LazyDataTableTheme({
    this.columnHeaderBorder,
    this.rowHeaderBorder,
    this.cellBorder,
    this.cornerBorder,
    this.columnHeaderColor,
    this.rowHeaderColor,
    this.cellColor,
    this.cornerColor,
  });

  /// [BoxBorder] for the column header.
  final BoxBorder? columnHeaderBorder;

  /// [BoxBorder] for the row header.
  final BoxBorder? rowHeaderBorder;

  /// [BoxBorder] for the cell.
  final BoxBorder? cellBorder;

  /// [BoxBorder] for the corner widget.
  final BoxBorder? cornerBorder;

  /// [Color] for the column header.
  final Color? columnHeaderColor;

  /// [Color] for the row header.
  final Color? rowHeaderColor;

  /// [Color] for the cell.
  final Color? cellColor;

  /// [Color] for the corner widget.
  final Color? cornerColor;
}

/// A custom synchronized scroll controller.
///
/// This controller stores all their attached [ScrollPosition] in a list,
/// and when given a notification via [processNotification], it will scroll
/// every ScrollPosition in that list to the same [offset].
class _CustomScrollController extends ScrollController {
  _CustomScrollController(TickerProvider provider) : super() {
    _ticker = provider.createTicker((_) {
      jumpTo(offset + _velocity);
      _velocity *= 0.9;
      if (_velocity < 0.1 && _velocity > -0.1) {
        _ticker.stop();
      }
    });
  }

  /// List of [ScrollPosition].
  List<ScrollPosition> _positions = [];

  /// The offset of the ScrollPositions.
  double offset = 0;

  /// Ticker to calculate the [_velocity].
  late Ticker _ticker;

  /// The velocity of the controller.
  /// The [_ticker] will tick while the velocity
  /// is not between -0.1 and 0.1.
  late double _velocity;

  /// Stores given [ScrollPosition] in the list and
  /// set the initial offset of that ScrollPosition.
  @override
  void attach(ScrollPosition position) {
    position.correctPixels(offset);
    _positions.add(position);
  }

  /// Removes given [ScrollPostion] from the list.
  @override
  void detach(ScrollPosition position) {
    _positions.remove(position);
  }

  /// Processes notification from one of the [ScrollPositions] in the list.
  void processNotification(ScrollNotification notification) {
    if (notification is ScrollUpdateNotification) {
      jumpTo(notification.metrics.pixels);
    }
  }

  /// Jumps every item in the list to the given [value],
  /// except the ones that are already at the correct offset.
  @override
  void jumpTo(double value) {
    if (value > _positions[0].maxScrollExtent) {
      offset = _positions[0].maxScrollExtent;
    } else if (value < 0) {
      offset = 0;
    } else {
      offset = value;
    }
    for (ScrollPosition position in _positions) {
      if (position.pixels != offset) {
        position.jumpTo(offset);
      }
    }
  }

  /// Jump to [offset] + [value].
  void jump(double value) {
    jumpTo(offset + value);
  }

  /// Set [_velocity] to new value.
  void setVelocity(double velocity) {
    if (_ticker.isActive) _ticker.stop();
    _velocity = velocity;
    _ticker.start();
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }
}


example/lib/simple_table.dart
import 'package:flutter/material.dart';
import 'package:lazy_data_table_plus/lazy_data_table_plus.dart';

class SimpleTable extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Simple table"),
      ),
      body: LazyDataTable(
        rows: 100,
        columns: 100,
        tableDimensions: LazyDataTableDimensions(
          cellHeight: 50,
          cellWidth: 100,
          columnHeaderHeight: 50,
          rowHeaderWidth: 75,
        ),
        tableTheme: LazyDataTableTheme(
          columnHeaderBorder: Border.all(color: Colors.black38),
          rowHeaderBorder: Border.all(color: Colors.black38),
          cellBorder: Border.all(color: Colors.black12),
          cornerBorder: Border.all(color: Colors.black38),
          columnHeaderColor: Colors.white60,
          rowHeaderColor: Colors.white60,
          cellColor: Colors.white,
          cornerColor: Colors.white38,
        ),
        columnHeaderBuilder: (i) => Center(child: Text("Column: ${i + 1}")),
        rowHeaderBuilder: (i) => Center(child: Text("Row: ${i + 1}")),
        dataCellBuilder: (i, j) => Center(child: Text("Cell: $i, $j")),
        cornerWidget: Center(child: Text("Corner")),
      ),
    );
  }
}
